### ISSUE TITLE
Error when trying to use behind an nginx reverse proxy

### ISSUE THREAD

**Issue by KnightWhoSaysNeeeowWumPing** [TandoorRecipes/recipes#2427]
### Issue

Hey,

when trying to setup TandoorRecipes to be used behind an nginx reverse proxy I get the *500 Internal Server Error*.  
My setup is as follows:  
Ubuntu Server on the host machine (192.168.178.20) with a pretty basic nginx setup that routes `192.168.178.20/recipes` to tandoor at a VM that also runs Ubuntu Server (192.168.178.157).  
On this VM tandoor is running via docker as seen in the attached config files.

A snippet from the nginx config on the host can be seen in the following:
```
location /recipes/ {
    proxy_pass http://192.168.178.157:8123/;
    proxy_redirect http://192.168.178.157:8123 http://192.168.178.20/recipes/;
}   
```

The problem is that I get the 500 error when I set `SCRIPT_NAME=/recipes` in the `.env` file no matter how I try to access it (via proxy at `192.168.178.20/recipes` as well as directly via `192.168.178.157:8123`).  
Once I comment out that line, I am able to access tandoor directly (via `192.168.178.157:8123`) just fine (as expected).  
But once I try to access it through the proxy (via `192.168.178.20/recipes`) the basic website is loading, but all external resources (*.css, *.js, etc.) fail to load.

What am I missing here?

Thank you and keep up the great work!

### Tandoor Version

"latest", so I assume 1.4.8

### OS Version

Ubuntu 22.04.2 LTS

### Setup

Docker / Docker-Compose

### Reverse Proxy

Others (please state below)

### Other

regular nginx (see text)

### Environment file

```shell
DEBUG=0
SQL_DEBUG=0

ALLOWED_HOSTS=*

SECRET_KEY=[REDACTED]

TIMEZONE=[REDACTED]

DB_ENGINE=django.db.backends.postgresql
POSTGRES_HOST=db_recipes
POSTGRES_PORT=5432
POSTGRES_USER=[REDACTED]
POSTGRES_PASSWORD=[REDACTED]
POSTGRES_DB=[REDACTED]

FRACTION_PREF_DEFAULT=0

COMMENT_PREF_DEFAULT=1

SHOPPING_MIN_AUTOSYNC_INTERVAL=5

SCRIPT_NAME=/recipes

GUNICORN_MEDIA=0

REVERSE_PROXY_AUTH=0
```

### Docker-Compose file

```shell
version: "3" 
services:
  db_recipes:
    restart: always
    image: postgres:11-alpine
    volumes:
      - ./postgresql:/var/lib/postgresql/data
    env_file:
      - ./.env

  web_recipes:
    restart: always
    image: vabene1111/recipes
    env_file:
      - ./.env
    volumes:
      - staticfiles:/opt/recipes/staticfiles
      - nginx_config:/opt/recipes/nginx/conf.d
      - ./mediafiles:/opt/recipes/mediafiles
    depends_on:
      - db_recipes

  nginx_recipes:
    image: nginx:mainline-alpine
    restart: always
    ports:
      - 8123:80
    env_file:
      - ./.env
    depends_on:
      - web_recipes
    volumes:
      - nginx_config:/etc/nginx/conf.d:ro
      - staticfiles:/static:ro
      - ./mediafiles:/media:ro

volumes:
  nginx_config:
  staticfiles:
```

### Relevant logs

```shell
[2023-04-19 20:24:51 +0200] [13] [ERROR] Error handling request /
Traceback (most recent call last):
  File "/opt/recipes/venv/lib/python3.10/site-packages/gunicorn/workers/gthread.py", line 271, in handle
    keepalive = self.handle_request(req, conn)
  File "/opt/recipes/venv/lib/python3.10/site-packages/gunicorn/workers/gthread.py", line 308, in handle_request
    resp, environ = wsgi.create(req, conn.sock, conn.client,
  File "/opt/recipes/venv/lib/python3.10/site-packages/gunicorn/http/wsgi.py", line 183, in create
    path_info = path_info.split(script_name, 1)[1]
IndexError: list index out of range
172.18.0.2 - - [19/Apr/2023:20:24:51 +0200] "GET / HTTP/1.0" 500 0 "-" "-"
```

**Comment by smilerz** [comment]
using subfolder requires settings the additional headers below.

```
location ^~ /recipes {
    proxy_set_header X-Script-Name /recipes;
    proxy_cookie_path / /recipes;
}
```

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
Thank you for your fast reply.

I tried setting these headers, but it seems as if the resources are still not loaded correctly, as can be seen in the browser console:
```
GEThttp://192.168.178.20/recipes/login
[HTTP/1.1 404 Not Found 44ms]

This page is in Quirks Mode. Page layout may be impacted. For Standards Mode use “<!DOCTYPE html>”.
login
GEThttp://192.168.178.20/static/themes/tandoor.min.c3693f28cf81.css
[HTTP/1.1 404 Not Found 12ms]

GEThttp://192.168.178.20/static/css/app.min.9a45a418e522.css
[HTTP/1.1 404 Not Found 18ms]

GEThttp://192.168.178.20/static/js/jquery-3.5.1.min.dc5e7f18c8d3.js
[HTTP/1.1 404 Not Found 11ms]

GEThttp://192.168.178.20/static/js/popper.min.8156384b48c7.js
[HTTP/1.1 404 Not Found 27ms]

GEThttp://192.168.178.20/static/js/bootstrap.min.575d3de8fa36.js
[HTTP/1.1 404 Not Found 29ms]

GEThttp://192.168.178.20/static/css/select2.min.9f54e6414f87.css
[HTTP/1.1 404 Not Found 15ms]

GEThttp://192.168.178.20/static/js/select2.min.0f64f3a3a0c6.js
[HTTP/1.1 404 Not Found 15ms]

GEThttp://192.168.178.20/static/css/select2-bootstrap.dcec0cc93fba.css
[HTTP/1.1 404 Not Found 20ms]

GEThttp://192.168.178.20/static/fontawesome/fontawesome_all.min.2acd020d0b21.css
[HTTP/1.1 404 Not Found 8ms]

GEThttp://192.168.178.20/static/themes/select2-bootstrap-theme.d007f71f4d38.css
[HTTP/1.1 404 Not Found 18ms]

GEThttp://192.168.178.20/static/assets/brand_logo.6ebe02bf6707.png
[HTTP/1.1 404 Not Found 5ms]

Loading failed for the <script> with source “http://192.168.178.20/static/js/jquery-3.5.1.min.dc5e7f18c8d3.js”. login:34:1
Loading failed for the <script> with source “http://192.168.178.20/static/js/popper.min.8156384b48c7.js”. login:36:1
Loading failed for the <script> with source “http://192.168.178.20/static/js/bootstrap.min.575d3de8fa36.js”. login:37:1
GEThttp://192.168.178.20/static/css/select2.min.9f54e6414f87.css
[HTTP/1.1 404 Not Found 5ms]

GEThttp://192.168.178.20/static/js/select2.min.0f64f3a3a0c6.js
[HTTP/1.1 404 Not Found 5ms]

Loading failed for the <script> with source “http://192.168.178.20/static/js/select2.min.0f64f3a3a0c6.js”. login:41:1
GEThttp://192.168.178.20/static/themes/select2-bootstrap-theme.d007f71f4d38.css
[HTTP/1.1 404 Not Found 5ms]

GEThttp://192.168.178.20/static/css/select2-bootstrap.dcec0cc93fba.css
[HTTP/1.1 404 Not Found 5ms]

Uncaught ReferenceError: $ is not defined
    <anonymous> http://192.168.178.20/recipes/login:49
login:49:9
GEThttp://192.168.178.20/static/fontawesome/fontawesome_all.min.2acd020d0b21.css
[HTTP/1.1 404 Not Found 5ms]

GEThttp://192.168.178.20/static/assets/brand_logo.6ebe02bf6707.png
[HTTP/1.1 404 Not Found 8ms]

GEThttp://192.168.178.20/static/assets/apple-touch-icon.0fcae5e85a72.png
[HTTP/1.1 404 Not Found 0ms]

GEThttp://192.168.178.20/static/assets/favicon.5382f9b2ece9.svg
```
*I tried uploading the screenshot, but GitHub wouldn't let me..*

**Comment by smilerz** [comment]
please share your entire nginx configuration

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
```
server {
   listen 80;

   location / {
        proxy_pass http://192.168.178.157:80/;
   }

   location /recipes/ {
        proxy_pass http://192.168.178.157:8123/;
	proxy_redirect http://192.168.178.157:8123 http://192.168.178.20/recipes/;
   }

   location ^~ /recipes {
	proxy_set_header X-Script-Name /recipes;
	proxy_cookie_path / /recipes;
   }
 }
```

**Comment by smilerz** [comment]
where you are you serving `/static` and `/media`?

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
`/static` (and I suppose this to be true for `/media` also) are getting served at `http://192.168.178.157:8123/static/...` as well as `http://192.168.178.20/recipes/static...` but according to the browser console, tandoori searches at `http://192.168.178.20/static/` where they are not served.

Do I need some kind of redirect statement for that within my nginx config, too?

**Comment by smilerz** [comment]
> `/static` (and I suppose this to be true for `/media` also) are getting served at `http://192.168.178.157:8123/static/...` as well as `http://192.168.178.20/recipes/static...` but according to the browser console, tandoori searches at `http://192.168.178.20/static/` where they are not served.
> 
> Do I need some kind of redirect statement for that within my nginx config, too?

I don't see a configuration for them in your nginx at all.

**Comment by smilerz** [comment]
No, just those two paths which only have static content need a web server.  Try clearing all local cache (shift-F5 on most browsers) may clear up any lingering issues.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
I cleared all cache from the browser but I still get redirected to `http://192.168.178.20/accounts/login/` instead of `http://192.168.178.20/recipes/accounts/login/` after trying to login.
(This of course yields a 403 error due to failed CSRF verification - as it should at this point.)

For reference, this is my current nginx config:
```
server {
   listen 80;

   location / {
        proxy_pass http://192.168.178.157:80/;
   }

   location /recipes/ {
        proxy_pass http://192.168.178.157:8123/;
	proxy_redirect http://192.168.178.157:8123 http://192.168.178.20/recipes/;
   }

   location ^~ /recipes {
	proxy_set_header X-Script-Name /recipes;
	proxy_cookie_path / /recipes;
   }

   location /static/ {
        proxy_pass http://192.168.178.20/recipes/static/;
   }

   location /media/ {
        proxy_pass http://192.168.178.20/recipes/media/;
   }
}
```

**Comment by smilerz** [comment]
The /recipe configs need to be in a single block.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
That resolves it, thank you very much!

To get rid of the CSRF error, I additionally added the following headers back to the config:
```
proxy_set_header Host $http_host;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-Proto $scheme;
```

I'm still not able to view existing recipes though (also getting redirected to `http://192.168.178.20/view/recipe/...` without the `/recipes/`) but at least the login is working.

I guess I'll give it another shot tomorrow.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
Still no luck... I'm being redirected to `http://192.168.178.20/view/recipe/...` instead of `http://192.168.178.20/recipes/view/recipe/...` when trying to view existing recipes.
When trying to access a recipe by manipulating the URL to include the `/recipes/`, I get the following error pop-up:
```
There was an error loading a resource! "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>404 Not Found</title>\n</head><body>\n<h1>Not Found</h1>\n<p>The requested URL was not found on this server.</p>\n<hr>\n<address>Apache/2.4.52 (Ubuntu) Server at 192.168.178.157 Port 80</address>\n</body></html>\n"
```
I also tried both, `$host` as well as `$http_host` fo the `Host` and `X-Forwarded-Host` headers.
Playing around with the `X-Forwarded-For` header was not successful either.

Can you or anyone else think of what might be missing?

**TL;DR:**
I'm using the docker image of tandoor on a virtual machine (IP: 192.168.178.157 - Port: 8123 - docker-compose and env see above).
On the host machine, I'm running nginx as a reverse proxy (IP: 192.168.178.20 - my current nginx config is as follows).

```
server {
   listen 80; 

   location / { 
        proxy_pass http://192.168.178.157:80/;
   }   

   location ^~ /recipes {
        proxy_set_header X-Script-Name /recipes;
        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_cookie_path / /recipes;
        proxy_pass http://192.168.178.157:8123;
        proxy_redirect http://192.168.178.157:8123 http://192.168.178.20/recipes;
   }

   location /static/ {
        proxy_pass http://192.168.178.20/recipes/static/;
   }

   location /media/ {
        proxy_pass http://192.168.178.20/recipes/media/;
   }
 }
```

My goal is to set up the reverse proxy to serve `http://192.168.178.157:80` when accessing `http://192.168.178.20` (which works as intended) while serving tandoor `http://192.168.178.157:8123` at `http://192.168.178.20/recipes`.

**Comment by KnightWhoSaysNeeeowWumPing** [comment]
Never mind - I forgot that I commented out `SCRIPT_NAME=/recipes` in my .env yesterday. :facepalm: 
It's working now.

Thanks again! You were a big help!
